packages/engine/src/worker/runner/javascript/data_ffi.rs

```diff
@@ -0,0 +1,119 @@
+
+/// C representation of Arrow array data nodes
+#[repr(C)]
+#[derive(Debug)]
+pub struct DataFfi<'s> {
+    pub len: usize,
+    pub null_count: usize,
+    pub n_buffers: u32,
+    pub buffer_ptrs: [*const u8; 2],
+    pub buffer_capacities: [usize; 2],
+    pub null_bits_ptr: *const u8,
+    pub null_bits_capacity: usize,
+    pub(crate) _phantom: std::marker::PhantomData<v8::Local<'s, ()>>,
+}
+
+impl<'s> DataFfi<'s> {
+    /// Translation of the old mv8::mv8_data_node_from_js from C++ to Rust
+    pub(crate) fn new_from_js(
+        scope: &mut v8::HandleScope<'s>,
+        data: v8::Local<'s, v8::Value>,
+    ) -> Result<DataFfi<'s>> {
+        // The names in this function are kept identical to help the review process
+        tracing::debug!("Calling data_node_from_js");
+        let obj = data
+            .to_object(scope)
+            .ok_or_else(|| Error::V8("Could not convert data from Value to Object".to_string()))?;
+
+        let len_value = get_obj_property(scope, obj, "len")?;
+        let len_num: v8::Local<'s, v8::Number> = len_value.try_into().map_err(|err| {
+            Error::V8(format!(
+                "Could not convert len_value from Value to Number: {err}"
+            ))
+        })?;
+
+        let null_count_value = get_obj_property(scope, obj, "null_count")?;
+        let null_count_num: v8::Local<'s, v8::Number> =
+            null_count_value.try_into().map_err(|err| {
+                Error::V8(format!(
+                    "Could not convert null_count_value from Value to Number: {err}"
+                ))
+            })?;
+
+        let buffers_value = get_obj_property(scope, obj, "buffers")?;
+        let buffers: v8::Local<'s, v8::Array> = buffers_value.try_into().map_err(|err| {
+            Error::V8(format!(
+                "Could not convert buffers from Value to Array: {err}"
+            ))
+        })?;
+        let n_buffers = buffers.length();
+        if n_buffers > 2 {
+            return Err(Error::V8(format!(
+                "Invalid buffers length ({}), expected no more than 2",
+                buffers.length()
+            )));
+        }
+
+        let mut buffer_ptrs = [std::ptr::null(); 2];
+        let mut buffer_capacities = [0; 2];
+        for i in 0..buffers.length() {
+            let buffer_value = buffers
+                .get_index(scope, i)
+                .ok_or_else(|| Error::V8(format!("Could not access index {i} on buffers")))?;
+            let buffer: v8::Local<'s, v8::ArrayBuffer> =
+                buffer_value.try_into().map_err(|err| {
+                    Error::V8(format!(
+                        "Could not convert buffer_value from Value to ArrayBuffer: {err}"
+                    ))
+                })?;
+            let contents = buffer.get_backing_store();
+            buffer_ptrs[i as usize] = contents
+                .data()
+                .map(NonNull::as_ptr)
+                .unwrap_or(std::ptr::null_mut()) as *mut u8;
```

Comment:
```suggestion
                .unwrap_or(std::ptr::null_mut())
                .cast();
```

Reply:
Addressed in https://github.com/hashintel/hash/pull/414/commits/bd720edae72a53c2e3f343851b4addfb7031d768